src/group_theory/pi_hom.lean

```diff
@@ -0,0 +1,232 @@
+/-
+Copyright (c) 2022 Jocchim Breitner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+import group_theory.subgroup.basic
+import tactic.group
+import group_theory.general_commutator
+import group_theory.order_of_element
+import data.finset.noncomm_prod
+
+/-!
+# TODO
+
+-/
+
+variables {G : Type*} [group G]
+
+lemma mul_eq_one_of_disjoint
+  {H₁ H₂ : subgroup G} (hdis : disjoint H₁ H₂) {x y : G} (hx : x ∈ H₁) (hy : y ∈ H₂)
+  (heq : x * y = 1) : x = 1 ∧ y = 1 :=
+begin
+  have : y = x⁻¹ := symm (inv_eq_iff_mul_eq_one.mpr heq),
+  subst this,
+  have hy := H₂.inv_mem_iff.mp hy,
+  have : x ∈ H₁ ⊓ H₂, by { simp, cc },
+  rw [hdis.eq_bot, subgroup.mem_bot] at this,
+  subst this,
+  simp
+end
+
+namespace pi_hom
+
+variables {I : Type*} [fintype I] {H : I → Type*} [∀ i, group (H i)] (ϕ : Π (i : I), H i →* G)
+
+variables (f g : Π (i : I), H i)
+
+-- A bit like `finset.noncomm_prodi`, but without the dependent Prop arg that makes things tricky
+
+/-- TODO -/
+noncomputable
+def fun_on (S : finset I) : G := (list.map (λ i, ϕ i (f i)) S.to_list).prod
```

Comment:
I think it's unprovable whether `fun_on ϕ f {i, j}` is `ϕ i (f i) * ϕ j (f j)` or `ϕ j (f j) * ϕ i (f i)`, right? If so, this API is a little weird.

Reply:
It's the same, but I gotta start with something, don't I?

I consider this internal to the construction. Should I mark it as `private`?

It's inspired by https://leanprover-community.github.io/mathlib_docs/data/finset/noncomm_prod.html#finset.noncomm_prod but without a proof obligation that mentions `S`.

I could `include hcomm` for this definition (and not just the proofs for it) if it makes it less odd?

Reply:
Hmm, maybe I can use `noncomm_prod` after all, if I hide the annoying proof argument in a definition like this `fun_on`, discharged via the `hcomm` assumption. Then I can simplify this stuff here a bit. Will try.

Reply:
Interesting, `noncomm_prod` is `noncomputable`. That’s a good reason to use that…

Reply:
I’ve ported it to use `noncomm_prod`. Benefits: Less to prove, and no `noncomputable`. 

Downside: The API for `noncomm_prof` insists using `insert a s` over `cons a s`. This requires `[decidable_eq I]`. The simp-setup of `finset` also rewrites `cons` to `insert`, but then lemmas like
```
lemma fun_on_insert_of_not_mem (S : finset I) (i : I) (h : i ∉ S) :
  fun_on ϕ f (insert i S) = ϕ i (f i) * fun_on ϕ f S :=
```
don’t work well as `simp` anymore, because the `h` is no longer mentioned on the LHS. Oh well.